### ISSUE TITLE
Incorrect CPU usage reported for Windows based processor group setup #1361

### ISSUE THREAD

**Issue by RajeshPutta87** [oshi/oshi#1373]
Hi Daniel,

This is a follow up to the issue reported as part of https://github.com/oshi/oshi/issues/1361. Still the issue is not yet solved. 

We have validated and could see the incorrect logical processor issue (#1360) is solved. We could see correct 128 logical processor count).

But the incorrect CPU usage issue which was logged as part of this issue #1361 is still not yet solved. CPU usage is detected on only half the cores. Can you please look into it.

I have attached jar file here. I have changed extension jar to txt for your reference. Please change extension to jar for your reference
oshi-core-5.1.0.txt

Please let me know in case you need any specific details.
[oshi-core-5.1.0.txt](https://github.com/oshi/oshi/files/5419800/oshi-core-5.1.0.txt)

**Comment by dbwiddis** [comment]
> Can you please look into it.

Unfortunately without access to a machine with more than 64 cores, it's very difficult.  I'm going to need some debugging information from the relevant machine (or any Windows machine with more than 64 cores).

Here's the summary of the situation:
* OSHI gets the processor ticks from the performance counters
* The performance counter instances are indexed by NUMA node and processor number
* The previous version of OSHI used the processor number from the bitmask, which we know (from the fixed "missing 64th processor" bug) went 0 to 63 for each of the two processor groups. This was based on [this description of processor groups](https://docs.microsoft.com/en-us/windows/win32/procthread/processor-groups):
> A logical processor is identified by its group number and its group-relative processor number. This is represented by a PROCESSOR_NUMBER structure. Numeric processor numbers used by legacy functions are group-relative.
* The new version that I thought would fix the problem indexed 0 to 31 for each of the NUMA nodes. This is [how performance counters are indexed](https://superuser.com/a/942391/181023).

As mentioned in #1361, there are several points of failure here and I need debugging information to determine the correct fix.  
* It is possible that the retrieval of the performance counters only gets information for the "current" processor group.  From the above link:
> Applications that manage, maintain, or display per-processor information for the entire system must be modified to support more than 64 logical processors. An example of such an application is Windows Task Manager, which displays the workload of each processor in the system.
In this situation, it's possible that Java can only see processors in its current group. The discussion in #785 is relevant here, as [Java does not support processor groups](https://github.com/oshi/oshi/issues/785#issuecomment-458271154). 
* It is possible that the numbering (NUMA,Processor) of the "perfmon" instances mismatches the numbering of the counters. This is what I thought I fixed, but I'm not completely sure.

To debug further, I need to inspect the output of the method `queryProcessorCounters()` in the `ProcessorInformation` class.   Is it possible for your customer to execute a Java program on their machine?  Executing this code would provide the needed output:
```
    public static void main(String[] args) {
        Pair<List<String>, Map<ProcessorTickCountProperty, List<Long>>> debug = queryProcessorCounters();
        List<String> instances = debug.getA();
        Map<ProcessorTickCountProperty, List<Long>> counters = debug.getB();
        for (int i = 0; i < instances.size(); i++) {
            System.out.format("%s: %d%n", instances.get(i),
                    counters.get(ProcessorTickCountProperty.PERCENTPROCESSORTIME).get(i));
        }
    }
```
As a possible proxy/backup, executing the following command in a command shell would be useful:
```
wmic path Win32_PerfRawData_Counters_ProcessorInformation get Name,PercentProcessorTime
```

**Comment by RajeshPutta87** [comment]
@dbwiddis Thanks for the update. I have added a utility class to gather details as per your ask. I have shared those details with my team to share them with customer to get the results. I will update you as i get the update from customer

Thanks
Rajesh

**Comment by dbwiddis** [comment]
I'm fairly confident that this has been fixed in release 5.3.2.  I get expected results on a test VM.  

I'll leave this open while you get more details from your customer, but recommend you update your software to the newest version and ask the customer to test it.

**Comment by RajeshPutta87** [comment]
Thanks Daniel. I will await observations from customer and share them with you. Sure i will pass on recommendations to upgrade our software to use OSHI 5.3.2 version to be checked at customer end

**Comment by dbwiddis** [comment]
Details of the changes:
* 5.3.0 and earlier had proper processor numbering by group (group 0, proc 0-63; group 1, proc 0-63) but used the processor number to combine with the NUMA node, meaning nodes 1 and 3 missed updating from the perfmon counters (looked for keys 1,32; 1,33, ... instead of actual keys 1,0; 1,1, ...).  Also the bitmask math lost processor #63 in both groups.
* 5.3.1 changed the processor number to be by NUMA node.   I think the released version should have given correct results for cpu usage, but would still be incorrect for processor numbering, very relevant for affinity calculations.  I'm not sure why this test failed for your customer, but it's possible copying code from the snapshot version missed something.
* 5.3.2 restored the original (correct/standard/affinity-based) processor numbering but correctly handled the numa node indexing which matches the perfmon counters.

**Comment by dbwiddis** [comment]
Closing this as I'm confident it's resolved. If not, feel free to re-open!

**Comment by RajeshPutta87** [comment]
Hi Daniel,

We have upgraded our software to leverage OSHI 5.3.2 version and got it checked at customer end. It didnot not solve the issue. Also please find the output of earlier ask

`SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".

SLF4J: Defaulting to no-operation (NOP) logger implementation

SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.

3,31: 345907205468750

3,30: 346159779687500

3,29: 346297879687500

3,28: 346114306250000

3,27: 346326344843750

3,26: 346241433593750

3,25: 346321892343750

3,24: 346241100625000

3,23: 346319175625000

3,22: 346247701406250

3,21: 346330600937500

3,20: 346227790625000

3,19: 346344962812500

3,18: 346189369531250

3,17: 346102590000000

3,16: 346153056562500

3,15: 346324457187500

3,14: 344824677187500

3,13: 346330755468750

3,12: 344799365781250

3,11: 346309167031250

3,10: 346196822656250

3,9: 346327743906250

3,8: 346198480781250

3,7: 346277567343750

3,6: 346144365625000

3,5: 346283860625000

3,4: 346186897656250

3,3: 346272785156250

3,2: 346186825156250

3,1: 346271073437500

3,0: 346148217187500

2,31: 346423571093750

2,30: 346302963750000

2,29: 346442454375000

2,28: 345714164531250

2,27: 345703236406250

2,26: 346319047968750

2,25: 346309784687500

2,24: 346308143750000

2,23: 346450505781250

2,22: 346242595781250

2,21: 346449878906250

2,20: 346320850312500

2,19: 346459061562500

2,18: 346324831562500

2,17: 346418131875000

2,16: 346290277656250

2,15: 346433641875000

2,14: 346231935937500

2,13: 346436542031250

2,12: 346242973906250

2,11: 346429737656250

2,10: 346206471718750

2,9: 346334293437500

2,8: 346250476875000

2,7: 346453490625000

2,6: 346249953906250

2,5: 346389217187500

2,4: 345719603750000

2,3: 346430210937500

2,2: 343131486406250

2,1: 346385309531250

2,0: 346111350625000

1,31: 345922508593750

1,30: 344389811562500

1,29: 344375285000000

1,28: 344384828906250

1,27: 344446842968750

1,26: 344331813906250

1,25: 344454133593750

1,24: 344353909062500

1,23: 344448889687500

1,22: 344193512343750

1,21: 344432347812500

1,20: 344372661406250

1,19: 344432789062500

1,18: 344358648906250

1,17: 344400978750000

1,16: 344353691875000

1,15: 344415261250000

1,14: 344348475312500

1,13: 344443700468750

1,12: 344347833593750

1,11: 344443636250000

1,10: 344332423593750

1,9: 344410835781250

1,8: 344345075156250

1,7: 344419082500000

1,6: 344338940625000

1,5: 344444235000000

1,4: 344300007343750

1,3: 344330704531250

1,2: 344377800781250

1,1: 344416146718750

1,0: 344413923281250

0,31: 346205181250000

0,30: 345202904218750

0,29: 345289041718750

0,28: 345190906718750

0,27: 345328899375000

0,26: 345151415625000

0,25: 345341633593750

0,24: 345150037343750

0,23: 345305400937500

0,22: 345106902031250

0,21: 345341014062500

0,20: 345149745468750

0,19: 345291845156250

0,18: 345119446406250

0,17: 345278144687500

0,16: 345136762500000

0,15: 345303133593750

0,14: 345150215937500

0,13: 345307471406250

0,12: 345151768750000

0,11: 345343890625000

0,10: 345075478281250

0,9: 345314858437500

0,8: 345126247656250

0,7: 345289476562500

0,6: 344841862500000

0,5: 345319065156250

0,4: 343493485937500

0,3: 345274944062500

0,2: 343373384062500

0,1: 345303498906250

0,0: 345058677656250

`

`E:\bg.downloads\AppDynamics>wmic path Win32_PerfRawData_Counters_ProcessorInformation get Name,PercentProcessorTime

Name PercentProcessorTime

_Total 337393200465087

3,_Total 338068251254882

3,31 337854182500000

3,30 338081560000000

3,29 338225939062500

3,28 338042448281250

3,27 338247328593750

3,26 338164136718750

3,25 338241586562500

3,24 338164001093750

3,23 338238878437500

3,22 338169319843750

3,21 338249358125000

3,20 338149431406250

3,19 338267823593750

3,18 338113061718750

3,17 338033630937500

3,16 338077539843750

3,15 338244267343750

3,14 336756026562500

3,13 338250375937500

3,12 336727556250000

3,11 338229447812500

3,10 338124347187500

3,9 338248397500000

3,8 338128805937500

3,7 338200608281250

3,6 338069568593750

3,5 338203662187500

3,4 338111414375000

3,3 338194123593750

3,2 338108558906250

3,1 338193312031250

3,0 338073340937500

2,_Total 338091481118164

2,31 338325750156250

2,30 338201406250000

2,29 338343002968750

2,28 337613630312500

2,27 337606122968750

2,26 338220425781250

2,25 338212629218750

2,24 338209606093750

2,23 338348863750000

2,22 338145502031250

2,21 338347916875000

2,20 338223155000000

2,19 338355173906250

2,18 338225755312500

2,17 338317167812500

2,16 338193052500000

2,15 338331748906250

2,14 338144100937500

2,13 338333198750000

2,12 338154356562500

2,11 338328542500000

2,10 338116329375000

2,9 338232795781250

2,8 338159831875000

2,7 338351764062500

2,6 338154261875000

2,5 338288875000000

2,4 337662491093750

2,3 338330553906250

2,2 335146086718750

2,1 338282361718750

2,0 338020935781250

1,_Total 336356302661132

1,31 337844943281250

1,30 336314722500000

1,29 336300272031250

1,28 336310239062500

1,27 336371075468750

1,26 336257640312500

1,25 336377995468750

1,24 336278757343750

1,23 336373133593750

1,22 336118095156250

1,21 336356926718750

1,20 336297202500000

1,19 336356573906250

1,18 336283663125000

1,17 336326131562500

1,16 336280099375000

1,15 336339761093750

1,14 336274379687500

1,13 336367982031250

1,12 336275910000000

1,11 336367837031250

1,10 336265022187500

1,9 336335794375000

1,8 336280560468750

1,7 336344076718750

1,6 336272338281250

1,5 336368101718750

1,4 336227902500000

1,3 336254438593750

1,2 336302099843750

1,1 336339669843750

1,0 336338339375000

0,_Total 337056766826171

0,31 338108991562500

0,30 337122045156250

0,29 337199472500000

0,28 337110302031250

0,27 337239718125000

0,26 337072016250000

0,25 337248799531250

0,24 337068076250000

0,23 337219696562500

0,22 337028012031250

0,21 337253755000000

0,20 337071932031250

0,19 337217619062500

0,18 337040510937500

0,17 337190244375000

0,16 337060378750000

0,15 337214831562500

0,14 337073071718750

0,13 337219395312500

0,12 337076396250000

0,11 337255380625000

0,10 336998282812500

0,9 337227237968750

0,8 337050195781250

0,7 337202729375000

0,6 336782148437500

0,5 337229880781250

0,4 335477985156250

0,3 337195854531250

0,2 335363240156250

0,1 337215553906250

0,0 336982783906250 `

**Comment by dbwiddis** [comment]
Thank you for the log file.  It validates what I was expecting (NUMA nodes numbered 0,0 thru 0,31; 1,0 thru 1,31, etc.  It also confirms that we have performance counter values for all 128 cores.  

But in that case, I can't see how OSHI is returning invalid data.  

How is the code using OSHI's values to sum up the totals displayed on the graph you showed?

**Comment by RajeshPutta87** [comment]
Hi Daniel,

Thanks for the update. This is how we are using OSHI

CentralProcessor cpuInfo = hal.getProcessor();
        int coreCount = cpuInfo.getPhysicalProcessorCount();
        int processorCount = cpuInfo.getLogicalProcessorCount();
        CentralProcessor.ProcessorIdentifier cpuId = cpuInfo.getProcessorIdentifier();
        Map<String, WindowsRawData.Cpu> cpusMap = ImmutableMap.of("CPU", WindowsRawData.Cpu.builder()
                .cpuVendor(cpuId.getVendor())
                .cpuName(cpuId.getName())
                .numCores(coreCount)
                .numLogicalProcessors(processorCount)
                .cpuMaxSpeedMhz(cpuId.getVendorFreq() / 1000000L)
                .build());
        long[] loadTicks = cpuInfo.getSystemCpuLoadTicks();
        log.debug("OSHI library output (SystemCpuLoadTicks): " + Arrays.toString(loadTicks));
        long idleTicks = loadTicks[CentralProcessor.TickType.IDLE.getIndex()];
        long stealTicks = loadTicks[CentralProcessor.TickType.STEAL.getIndex()];
        long systemTicks = loadTicks[CentralProcessor.TickType.SYSTEM.getIndex()]
                + loadTicks[CentralProcessor.TickType.IRQ.getIndex()]
                + loadTicks[CentralProcessor.TickType.SOFTIRQ.getIndex()];
        long userTicks = loadTicks[CentralProcessor.TickType.USER.getIndex()];
        long timeStamp = idleTicks + stealTicks + systemTicks + userTicks;
    Map<String, WindowsRawData.CpuPerf> cpusPerfMap = ImmutableMap.of("_Total", WindowsRawData.CpuPerf.builder()
                .timestamp(timeStamp)
                .idleTicks(idleTicks)
                .stealTicks(stealTicks)
                .systemTicks(systemTicks)
                .userTicks(userTicks)
                .build());
        rawData.setCpus(cpusMap);
        rawData.setCpusPerf(cpusPerfMap);
        log.debug("OSHI library output (cpus): " + rawData.getCpus());
        log.debug("OSHI library output (cpusPerf): " + rawData.getCpusPerf());

Reporting of data to our UI is at every 1 minute interval. By default we sample the data every 30 seconds. So we collect 2 samples in a minute using OSHI. For every minute data reporting we send data as shown here ->

Metric ID followed by 
sum of samples in a minute, 
number of samples collected in a minute
min of samples collected in a minute
max of samples collected in a minute
what is the current value (last reported sample in that minute)

<metric id='446',  value[sum=18, count=2, min=9, max=9, current=9]>

Thanks
Rajesh

**Comment by dbwiddis** [comment]
The debug log output for `log.debug("OSHI library output (SystemCpuLoadTicks): " + Arrays.toString(loadTicks));` would be very useful.

**Comment by RajeshPutta87** [comment]
Thanks Daniel. I will get those details and keep you updated.

Thanks
Rajesh

**Comment by dbwiddis** [comment]
Oh, wait.... with the original "missing processor #63" bug, I got focused in on the per-processor statistics.  But now looking again you're just collecting total system statistics.

That pulls from the "Processor" counter which is not Processor-group aware.  This explains the problem we're seeing and confirms the bug.

I need to switch that to pull from the "ProcessorInformation" counter.

In the meantime, you could just use the per-processor ticks and sum them up, as a workaround.

**Comment by dbwiddis** [comment]
No need to pull that debug log line.

**Comment by RajeshPutta87** [comment]
Hi Daniel,

Thanks for the update. I am slightly confused. Can you please confirm what changes you want me to make to earlier provided code snippet until you pull from the "ProcessorInformation" counter

Thanks
Rajesh

**Comment by dbwiddis** [comment]
You do not *need* to make any changes right now.  

If you want to work around this current incompatibility as a short-term workaround, you could pull the values from `getProcessorCpuLoadTicks()` and add up all of them.  That's essentially the change I'll be making in OSHI anyway.

**Comment by RajeshPutta87** [comment]
Thanks Daniel. I will wait for a OSHI build with a fix instead of me making any temporary changes, so that it will help us to achieve consistency with OSHI future builds. 

Thanks
Rajesh

**Comment by dbwiddis** [comment]
I've just released 5.3.5.

**Comment by RajeshPutta87** [comment]
Thanks Daniel

Regards
Rajesh
